### ISSUE TITLE
Documentation is missing screenshots.

### ISSUE THREAD

**Issue by edthedev** [kivy/kivy#710]
For this project to gain traction, there need to be screenshots for every page that has an example application. Especially the ones that us the .kv files.

I'm willing to contribute screenshots, if someone will point me to where to submit them.

**Comment by merriam** [comment]
This does sound like a good idea.   I could see rewriting doc/build/html/gettingstarted/examples.html to have nice screenshots and better descriptions.   Should I go ahead and do it?

Just kidding, I'll just put in a Pull Request rather than asking permission. ;)

**Comment by merriam** [comment]
OK, I'm getting closer to this one.   If anyone has sensibilties about the layout, please let me know now.

The idea is to have a gallery of screenshots, with descriptions pulled from the example docstrings,   
There are two screenshots showing how that looks.  Clicking on any screenshot goes to a detail
page that linky's the doc string and puts the source code for the example and related files on that page.


![screen shot detail3](https://cloud.githubusercontent.com/assets/159998/6123771/74a8e46a-b0bb-11e4-9d94-91ac70374fbc.png)
![screen shot detail2](https://cloud.githubusercontent.com/assets/159998/6123774/74c3b5e2-b0bb-11e4-9f3c-24e092e873a8.png)
![screen shot detail](https://cloud.githubusercontent.com/assets/159998/6123772/74c2ea36-b0bb-11e4-8820-028a4e955064.png)
![screen shot gallery2](https://cloud.githubusercontent.com/assets/159998/6123773/74c33946-b0bb-11e4-8357-4f843bab5513.png)
![screen shot gallery](https://cloud.githubusercontent.com/assets/159998/6123775/74c77ca4-b0bb-11e4-90a3-251d655bf128.png)

**Comment by picibucor** [comment]
Great, I love it :+1:

**Comment by rubik** [comment]
Awesome work!

**Comment by merriam** [comment]
There is still some work to be done on this issue:

* Document (add docstring) and upload screenshots for another 30 examples
* Update contributions/other pages to explain how documentation is generated
* Update or replace the GettingStarted / Examples page
* Automatically filter and link to major Kivy APIs based on example imports.
* Don't trigger rebuilding all examples on every make.

I'm willing to do the actual work, but no longer willing to deal with the PR cycle.   If someone is willing to handling the objections on each PR, I can either pass them the code or just let them work on it after the intial pull request.   For example, PR #3027, PR #3023, and PR#3029 move towards this issue, but require soothing to get merged.  I'm looking for someone who likes being the buffer.

**Comment by picibucor** [comment]
I really love our new gallery. Thx for it. The only way it could be improved is to change the order of examples. My personal view is that the very first example shouldn't be a 3d rendering with openGL....
I saw that the galery is in alphabetical order, and generated with sphinx. I'm not experienced with sphinx. My question: Is renaming our examples an easy way to reorganize the gallery? 
Under renaming I mean for example: 
instead of "app_with_build.py"  putting a number in front of it : "1_app_with_build.py"

**Comment by merriam** [comment]
My first version actually did that, naming screenshots like
"01__path__file__ext.py".  It's very doable, changing existing file names
and such.  Unfortunately, I'm no longer directly making changes to kivy:
 it frustrates me that every pull request comes back with requests for
useless changes.   I you care enough, I'll make the change and let you
fight with getting it accepted.

**Comment by picibucor** [comment]
Are these only PEP8 - related issues?

**Comment by merriam** [comment]
The issues feel more conflicts about appreciation and wanting
conversations.   Any PEP8 issues get a bounce, and reviewers are upset that
I often use GitHub's Mac tool which fails on commit hooks.  Other bounces
include odd things like too much space in the github archive, possible
experiments someone else will be running, and so on.   I currently have
four open PRs with one 'closed' but still in progress.   At least one PR
has been resubmitted several times, never on substantive changes.   I try
to not to have more than five PRs open, as it starts to cause more
management than coding.

Currently, I'm just going through the rest of the examples and will submit
a single PR to close most of the remaining issues in the examples
directory.

For the screenshots, ordering, how about:
- screenshots with 'ordering tags' are always first.
- 'ordering tags' are optional tags before the directory name and can only
be digits.
- This adds the restriction that an untagged screenshot may not come from a
directory named only with digits, but
  that's acceptable.
- The changes are changing the pattern to:
      pattern = re.compile(r'^((.+)__(.+)__([^-]+))\.png')
      also, adjust the group numbers, set a ['order_tag'] in the fileinfo,
and change the sort line to use order_tag first.
- What screenshot order did you have in mind?   Would a grid of thumbnails
be a better solution?

Charles

**Comment by picibucor** [comment]
:+1: 

And I like the thumbnail grid as well.

As a first idea I would suggest:
1. Application example using build() + return
2. Gallery of Examples » Application built from a .kv file
3. Application from a .kv in a Template Directory
4.Showcase of Kivy Features
5. Kivy Catalog
6. Camera Example
7. Lines Extended Demo
8. Line (SmoothLine) Experiment
9. Circle Example
10. Widget animation
11. Canvas stress
12. Rotation Example
13. Bezier Example
14. Repeat Texture on Resize
15. Texture Wrapping and Coordinates Example
16. Multitexture Example
17. Stencil demo
18. Shuffled Camera Feed Puzzle
19. Mesh test
20. Tesselate Demonstration
21. Suite of Application Builders
22. Basic Picture Viewer
23. Touch Tracer Line Drawing Demonstration
24. 3D Rotating Monkey Head
24. Multistroke Recognition Database Demonstration

**Comment by dessant** [comment]
I think it would be also good to auto-generate screenshots with kivy's screenshot method, without transparent borders and window decorations, as those elements feel outdated in 1-2 years and they are not representative because kivy is cross-platform.

**Comment by merriam** [comment]
In general, generating the screenshots would be better for the reasons
dessant mentions, as well as for extra unit tests.  The currently generated
screenshots are usually 'further into the app', meaning some user input
should already have taken place.   To automate the screenshot of an app
that starts with a blank screen is an interesting problem; I'm not clear
what's used for event replaying.   So, a good suggestion but not yet
practical.

**Comment by merriam** [comment]
I think I see what you are doing:  trying to make a progression of
applications that guide a new programmer through the library.  Less like a
tutorial and more like a book broken up by chapters?

So, how about adding a hierarchy and some 'just read me' text.  That is,
slice your list something like this:

Building Applications
    Readme note on purpose of examples and pointing to documentation
    1. Application example using build() + return
    2. Gallery of Examples » Application built from a .kv file
    3. Application from a .kv in a Template Directory
Playgrounds
    Readme note on purpose of examples and pointing to documentation
    4.Showcase of Kivy Features
    5. Kivy Catalog
Graphics
    Readme note on purpose of examples and pointing to documentation
   6. Camera Example
   7. Lines Extended Demo
   8. Line (SmoothLine) Experiment
   9. Circle Example
   10. Widget animation
   11. Canvas stress
   12. Rotation Example
   13. Bezier Example
   14. Repeat Texture on Resize
   15. Texture Wrapping and Coordinates Example
   16. Multitexture Example
   17. Stencil demo
   18. Shuffled Camera Feed Puzzle
   19. Mesh test
   20. Tesselate Demonstration
21. Suite of Application Builders
22. Basic Picture Viewer
23. Touch Tracer Line Drawing Demonstration
24. 3D Rotating Monkey Head
24. Multistroke Recognition Database Demonstration

Is this where you are heading?

**Comment by picibucor** [comment]
Yes. Definitely. It should have a "learning-curve". But I wouldn't name it as chapters of a book. I would name it what it is: "learning by examples". Every example in the same "chapter" is building on the previous example.

I've learned a lot playing around with some small basic examples. (I think it is not just me.) And that is why I like the matplotlib gallery as well. It allows you to see the very basic elements from you can build up your own plot / report / pdf / whatever. 
As a newbie with knowledge in python I would really like to experiment with small basic examples. And it is great to have something on the homepage to enhance these kind of interactions...

**Comment by victor-rene** [comment]
@picibucor @merriam 
As a newbie I second the two above posts. I best learned by copying examples (and asking questions on IRC). At the end of  the pong tutorial, it's mentioned that you should try examples and open the catalog (in "where to go from there?" section).

Providing a logical structure for a kind of crash course, will be a very nice addition to increase the potential for new "core" contributors. Anything done to ease the learning curve in this sense is probably a big bonus on the long term (cause I feel we lack man power). :+1:

**Comment by picibucor** [comment]
I think we need some more basic examples. SoI've added a new container-example. It is ment to be in the first section, after the playgrounds...
You could find it in my repo:
https://github.com/picibucor/kivy/tree/master/examples/container

**Comment by picibucor** [comment]
my first PR: #3117

**Comment by merriam** [comment]
Cool!

I like the idea of some examples showing, with explicit comments, how the
properties work and relate to the .kv language.

Good work!

Charles

**Comment by merriam** [comment]
I added the guided approach to looking at .kv files, but first down in Kivy Catalog.   The program example/demo/kivycatalog/main.py goes through a bunch of examples using .kv language files.  I added a category spinner so that examples can be found and paged through a little more comfortably:

This should show you what I mean:
![screen shot 2015-03-09 at 12 02 16 pm](https://cloud.githubusercontent.com/assets/159998/6562220/732eb782-c654-11e4-95d9-4fdde0fc51d2.png)

It's off in PR #3112, which won't get merged.   Someone could pull that section and resubmit, though I expect to drastically expand the number of examples in kivy_catalog over the next few days.  Might be better to wait.

**Comment by akshayaurora** [comment]
@merriam 
 - Make atomic commits. 
 - Don't bring in fixes that are not related to the issue your pr solves fix.
 - We have a commit hook. Use it. 
We have even school kids who are able to do this so I believe even you can too.
Simply put if any one can not follow the guidelines then we can't merge their pr.

**Comment by FeralBytes** [comment]
@merriam Do you need help chunking up PR #3112 to get it merged? Or are any of these: PR #3027, PR #3023, and PR #3029; higher priority?
@akshayaurora I will help him make better commits that will follow Kivy's guidelines.

**Comment by picibucor** [comment]
found a project with great documentation:
http://scikit-learn.org/stable/index.html

**Comment by FeralBytes** [comment]
@picibucor can you elaborate? What makes it so excellent? I did look at it, and it does have a lot of links for help. But what specifically do you see as excellent?

**Comment by dessant** [comment]
@FeralBytes , @picibucor: some stuff from https://github.com/kivy/kivy/pull/3112 was already taken out and pulled separately, it would be great to also pr the refactored screenshots, if someone finds the time.

**Comment by picibucor** [comment]
@FeralBytes and others

I recommended matplotlib/gallery because it is highly visual, the diagrams in the examples are well-separated in the code, and they are mostly pretty straightforward.

By matplotlib not only the examples are easy to understand, but your problem as well: You want to visualise a given dataset.

How to do it?
0. Have a set of data
1. Choose a type of diagram from the gallery
2. Reorganise your dataset as it is in the example
3. Copy-paste the chosen example, and replace some variable-names to your own dataset-variable name
4. Optionally you could have several types of diagrams on a plot, a layout of several plots, edit ticks/labels/layout, add button/slider/... (The more it is used in general, the easier it is to understand)

The examples in the matplotlib/gallery are parallel. They are only categorised, because the examples are not blocks built on each other. You could choose and use one without knowing about the others. 

Why is scikit-learn is a good example?

The problem kivy is solving is more complex. There are certain building blocks that are essential to know, blocks that are building on other blocks, and there are parallel parts as well.
I was searching for a documentation that manages this kind of complexity. This is where we scikit-learn comes in. Machine learning is a very complex topic as well. 
Please check this: http://scikit-learn.org/stable/tutorial/machine_learning_map/index.html

There are parallel parts, which contains other parallel parts or blocks built on other blocks. And we could use a similar kind of structure by the kivy/gallery_of_examples as well.

For example: 
![image](https://cloud.githubusercontent.com/assets/1676457/7102608/ba2ac824-e085-11e4-8632-03b6ee748587.png)

Maybe having such pictures showing the structure of what you are learning at the moment /where you are in the learning curve  would be also a good idea.

Other things I like by scikit-learn:

--> code-text-pic-ratio : 
Many pictures, small amount of texts and atomic examples. The longer the texts and/or code-examples are, the more motivation you need to sit down in front of it (even if you know the half of it). 
Watch the following video on the topic: https://www.youtube.com/watch?v=fqUSjHjIEFg .

I recommend using atomic examples. By a max. 8-12 line code your motivational barrier will low enough. If an example is longer than that, it could be chunked up (as the tutorials) and the differences between versions could be color-code.

--> large examples are hidden "behind" diagrams
Large examples could be marked somehow, signalling: "Extra motivation is needed". 

-->examples are downloadable
easier to play around

Basically that was it... Thx for reading.

**Comment by victor-rene** [comment]
@picibucor :+1:

Kivy is big
---
After a while, some people might take for granted that you know some of the basics, because they are written in the docs (or examples, etc...) but the truth is until you actually have to use them, you don't learn even if you have read about it. Then you are told, you were lazy :-(.

Let's say we had a map like the picture above, it would be slightly easier to check that you didn't miss something important during learning. As is, it's all text, code and experience. Nothing like a crash course.

## Better examples (with pics)
Another reason, is also that some common assumptions are being broken.
One good example is ScrollView. The [online doc](http://kivy.org/docs/api-kivy.uix.scrollview.html) covers a case that is not the most common one. A ScrollableLabel.
* Wiki: https://github.com/kivy/kivy/wiki/Scrollable-Label
* gist: https://gist.github.com/tshirtman/c8ed74df889c8452afb9

Providing examples that are actually relevant to day-to-day production code (possibly with relevant pictures) is a plus.

## Class hierarchy (maybe)
Maybe, it's not that hard to create a schema with the class hierarchies. Or you can also expand the doc with that class hierarchy in plain text like the MSDN online documentation. That would be a starting point even if not perfect.

## The problem with non-automated work (pictures)
Schemas, while quick to create often require manual intervention. Considering the steady pace of change of Kivy, it might be better to avoid such work if it cannot be automated. Because frankly, nobody will have time to update them (depends on frequency though). But I still like the ability to go through tutorials and example from a visual map.

__NOTE__
Off-topic. Someone from the core must define the perimeter of examples beyond the online doc, tutorials, including examples and wiki articles that constitute the real basics. Lastly, if anyone is interested I am building a list of basic issues and improvements on my [blog](http://victor-rene.com/software-eng).

**Comment by picibucor** [comment]
I made a gist for generating screenshots (as an idea):
https://gist.github.com/picibucor/fa44131c5b75c9b00230

Does ImageGrab run on other OS-s as well?

**Comment by picibucor** [comment]
@dessant :
Some screenshots added: #3374

**Comment by akshayaurora** [comment]
Some of the parts if this issue have been fixed. This issue is getting to big to follow.
Opening a new one to start fresh on this.
